sql/expression/function/time.go

```diff
@@ -1222,28 +1224,123 @@ func (c CurrTime) WithChildren(children ...sql.Expression) (sql.Expression, erro
 	return NoArgFuncWithChildren(c, children)
 }
 
+const maxCurrTimestampPrecision = 6
+
 type CurrTimestamp struct {
-	NoArgFunc
+	args []sql.Expression
 }
 
-var _ sql.FunctionExpression = CurrTimestamp{}
+var _ sql.FunctionExpression = (*CurrTimestamp)(nil)
 
-func NewCurrTimestamp() sql.Expression {
-	return CurrTimestamp{
-		NoArgFunc: NoArgFunc{"current_timestamp", sql.Datetime},
-	}
+func NewCurrTimestamp(args ...sql.Expression) (sql.Expression, error) {
+	return &CurrTimestamp{args}, nil
 }
 
 func currDatetimeLogic(ctx *sql.Context, _ sql.Row) (interface{}, error) {
 	return ctx.QueryTime(), nil
 }
 
-// Eval implements sql.Expression
-func (c CurrTimestamp) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
-	return currDatetimeLogic(ctx, row)
+func (c *CurrTimestamp) FunctionName() string {
+	return "current_timestamp"
 }
 
-// WithChildren implements sql.Expression
-func (c CurrTimestamp) WithChildren(children ...sql.Expression) (sql.Expression, error) {
-	return NoArgFuncWithChildren(c, children)
+func (c *CurrTimestamp) String() string {
+	return fmt.Sprintf("CURRENT_TIMESTAMP(%s)", c.args[0].String())
+}
+
+func (c *CurrTimestamp) Type() sql.Type { return sql.Datetime }
+
+func (c *CurrTimestamp) IsNullable() bool {
+	for _, arg := range c.args {
+		if arg.IsNullable() {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *CurrTimestamp) Resolved() bool {
+	for _, arg := range c.args {
+		if !arg.Resolved() {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *CurrTimestamp) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 0 && len(children) != 1 {
```

Comment:
Should check against the current length of the child expression slice.

The reason we do this check is to guard against bugs in the analyzer, and it should never call WithChildren with a different number of child expression than we already have.